slant-cgi.8
.\"	$Id$
.\"
.\" Copyright (c) 2018 Kristaps Dzonsons <kristaps@bsd.lv>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt SLANT-CGI 8
.Os
.Sh NAME
.Nm slant-cgi
.Nd CGI program returning slant data
.Sh SYNOPSIS
.Nm slant-cgi
.Sh DESCRIPTION
The
.Nm
utility produces system statistics as recorded by
.Xr slant-collectd 8 .
It interfaces with the database by default in
.Pa /var/www/data/slant.db .
.Pp
The CGI program accepts HTTP GET requests for the
.Pa /index.json
resource, which has the aliases
.Pa /index ,
.Pa / ,
or the empty request.
Other resources return an HTTP code 404.
Non-GET request return an HTTP code 405.
Other (non-200) codes are possible and follow standard definitions.
.Pp
On success,
.Nm
returns an HTTP code 200 and a valid JSON document.
.Pp
The document consists of the following.
In this description, integers
.Pq Li int
are serialised from 64-bit signed numbers.
Real-valued numbers
.Pq Li real
are serialised from double-precision numbers.
.Bd -literal
{   version: "x.y.z",
  timestamp: int,
     system: { system },
       qmin: [ records... ],
        min: [ records... ],
       hour: [ records... ],
        day: [ records... ],
       week: [ records... ],
       year: [ records... ] 
}
.Ed
.Pp
The
.Li version
is the software suite version.
This is informational.
.Pp
The
.Li system
object consists of system information:
.Bd -literal
{      boot: int,
    machine: string,
  osversion: string,
  osrelease: string,
    sysname: string
}
.Ed
.Pp
It consists only of
.Li boot ,
the UNIX epoch value for when the system was last booted.
The
.Li machine ,
.Li osversion ,
.Li osrelease ,
and
.Li sysname
mirror the return values of
.Xr uname 1 .
.Pp
The remaining values are the possibly-empty sets of records accumulated
over a given interval of time in quarter-minute quanta.
So each
.Li qmin
record consists of exactly one entry; each
.Li min
record consists of at least one and at most four entries, one for each
quarter-minute; each
.Li hour
has at most 240, etc.
.Pp
There are finite records (except for yearly ones) bound to a circular
buffer per interval.
When a record is filled, such as with four quarter-minute entries into a
minute record, the next entry creates a new record with one entry while
the oldest is discarded.
.Pp
The circular buffer is large enough for a reasonable glimpse into the
past, with emphasis placed on recent data: 40 quarter-minute records (10
minutes), 300 minute records (5 hours), 96 hourly (5 days), 28 daily (4
weeks), 104 weekly (two years), endless yearly entries.
.Pp
Each record consists of the following:
.Bd -literal
{    ctime: int,
   entries: int,
       cpu: real,
       mem: real,
     netrx: int,
     nettx: int,
  discread: int,
 discwrite: int,
    nprocs: real,
    rprocs: real,
    nfiles: real,
  interval: int,
        id: int 
}
.Ed
.Pp
The fields are defined as follows:
.Bl -tag -width Ds
.It Li ctime
a UNIX epoch of when the record was started (first entry)
.It Li entries
the number of entries, so front-end displays will need to average all
data by this number
.It Li cpu
average processor utilisation over all processing unit
.It Li mem
active memory over all available memory
.It Li netrx
bytes received per second over all interfaces
.It Li nettx
bytes transmitted per second over all interfaces
.It Li discread
bytes read per second over all configured devices
.It Li discwrite
bytes written per second over all configured devices
.It Li nprocs
number of running processes over all possible processes
.It Li rprocs
number of configured processes running over total configured
.It Li nfiles
number of open files over all possible open files
.It Li interval
the type of interval starting with zero for quarter-minute, one fo 
minute, etc.
.It Li id
a unique record identifier
.El
.\" The following requests should be uncommented and used where appropriate.
.\" .Sh CONTEXT
.\" For section 9 functions only.
.\" .Sh RETURN VALUES
.\" For sections 2, 3, and 9 function return values only.
.\" .Sh ENVIRONMENT
.\" For sections 1, 6, 7, and 8 only.
.\" .Sh FILES
.\" .Sh EXIT STATUS
.\" For sections 1, 6, and 8 only.
.\" .Sh EXAMPLES
.\" .Sh DIAGNOSTICS
.\" For sections 1, 4, 6, 7, 8, and 9 printf/stderr messages only.
.\" .Sh ERRORS
.\" For sections 2, 3, 4, and 9 errno settings only.
.Sh SEE ALSO
.Xr slant 1 ,
.Xr slant-collectd 8
.\" .Sh STANDARDS
.\" .Sh HISTORY
.\" .Sh AUTHORS
.\" .Sh CAVEATS
.\" .Sh BUGS


slant-cgi.c
/*	$Id$ */
/*
 * Copyright (c) 2018 Kristaps Dzonsons <kristaps@bsd.lv>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "config.h"

#if HAVE_SYS_QUEUE
# include <sys/queue.h>
#endif

#include <inttypes.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <kcgi.h>
#include <kcgijson.h>
#include <ksql.h>

#include "params.h"
#include "extern.h"
#include "db.h"
#include "json.h"

enum	page {
	PAGE_INDEX,
	PAGE__MAX
};

static const char *const pages[PAGE__MAX] = {
	"index", /* PAGE_INDEX */
};

/*
 * Fill out generic headers then start the HTTP document body (no more
 * headers after this point!)
 */
static void
http_open(struct kreq *r, enum khttp code)
{

	khttp_head(r, kresps[KRESP_STATUS], 
		"%s", khttps[code]);
	khttp_head(r, kresps[KRESP_CONTENT_TYPE], 
		"%s", kmimetypes[r->mime]);
	khttp_body(r);
}

static void
sendindex(struct kreq *r, 
	const struct system *sys, const struct record_q *q)
{
	struct kjsonreq	 req;
	const struct record *rr;

	http_open(r, KHTTP_200);
	kjson_open(&req, r);
	kjson_obj_open(&req);

	kjson_putstringp(&req, "version", VERSION);
	kjson_putintp(&req, "timestamp", time(NULL));

	json_system_obj(&req, sys);

	kjson_arrayp_open(&req, "qmin");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_byqmin == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);
	kjson_arrayp_open(&req, "min");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_bymin == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);
	kjson_arrayp_open(&req, "hour");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_byhour == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);
	kjson_arrayp_open(&req, "day");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_byday == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);
	kjson_arrayp_open(&req, "week");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_byweek == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);
	kjson_arrayp_open(&req, "year");
	TAILQ_FOREACH(rr, q, _entries)
		if (INTERVAL_byyear == rr->interval) {
			kjson_obj_open(&req);
			json_record_data(&req, rr);
			kjson_obj_close(&req);
		}
	kjson_array_close(&req);

	kjson_obj_close(&req);
	kjson_close(&req);
}

int
main(void)
{
	struct kreq	 r;
	enum kcgi_err	 er;
	struct record_q	*rq;
	struct system	*sys;

#if HAVE_PLEDGE
	if (-1 == pledge("stdio rpath "
	    "cpath wpath flock fattr proc", NULL)) {
		kutil_warn(NULL, NULL, "pledge");
		return EXIT_FAILURE;
	}
#endif

	er = khttp_parsex(&r, ksuffixmap,
             kmimetypes, KMIME__MAX, NULL, 0,
             pages, PAGE__MAX, KMIME_APP_JSON,
             PAGE_INDEX, NULL, NULL, 0, NULL);

	if (KCGI_OK != er) {
		kutil_warnx(NULL, NULL, "%s", kcgi_strerror(er));
		return EXIT_FAILURE;
	}

	/*
	 * Front line of defence: make sure we're a proper method, make
	 * sure we're a page, make sure we're a JSON file.
	 */

	if (KMETHOD_GET != r.method) {
		http_open(&r, KHTTP_405);
		khttp_free(&r);
		return EXIT_SUCCESS;
	} else if (PAGE__MAX == r.page || 
	           KMIME_APP_JSON != r.mime) {
		http_open(&r, KHTTP_404);
		khttp_free(&r);
		return EXIT_SUCCESS;
	}

	if (NULL == (r.arg = db_open(DBFILE))) {
		khttp_free(&r);
		return EXIT_SUCCESS;
	}

#if HAVE_PLEDGE
	if (-1 == pledge("stdio", NULL)) {
		kutil_warn(NULL, NULL, "pledge");
		db_close(r.arg);
		khttp_free(&r);
		return EXIT_FAILURE;
	}
#endif

	db_role(r.arg, ROLE_consume);

	rq = db_record_list_lister(r.arg);
	sys = db_system_get_id(r.arg, 1);

	sendindex(&r, sys, rq);

	db_system_free(sys);
	db_record_freeq(rq);

	db_close(r.arg);
	khttp_free(&r);

	return EXIT_SUCCESS;
}
